spring-core/src/main/java/org/springframework/core/retry/RetryCallback.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2002-2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.core.retry;
+
+/**
+ * Callback interface for a retryable piece of code. Used in conjunction with {@link RetryOperations}.
+ *
+ * @author Mahmoud Ben Hassine
+ * @since 7.0
+ * @param <R> the type of the result
+ * @see RetryOperations
+ */
+public interface RetryCallback<R> {
+
+	/**
+	 * Method to execute and retry if needed.
+	 * @return the result of the callback
+	 * @throws Throwable if an error occurs during the execution of the callback
+	 */
+	R run() throws Throwable;
+
+	/**
+	 * A unique logical name for this callback to distinguish retries around
+	 * business operations.
+	 * @return the name of the callback
+	 */
+	String getName();
```

Comment:
Not requirement, but I wonder if we can use a `default` convenient implementation here to the `getClass().getName()`?
Or generated lambda class name is not very useful?

Reply:
We have `ClassUtils.isLambdaClass()`, so we might be able to use that and base the name of the callback on the _enclosing class_ of the lambda.

But, TBH, I'm not sure what the enclosing class of a lambda is. I'd have to experiment with that.

Reply:
After our discussion, there is no need to introduce another interface.

> Not requirement, but I wonder if we can use a default convenient implementation here to the getClass().getName()?

I agree, @artembilan . I have used that as a default implementation for now. But I will let the team decide if we should return null and adapt the log messages accordingly.

Reply:
Spring Retry passes the callback itself to the listener methods (just to get the label), which is not very convenient in my opinion. But I agree, the `RetryExecution` object can be enriched with the callback name to give some context about the execution if needed.